### ISSUE TITLE
use_flash_attn and use_liger are read only by SFT's text path, yet validated on every task and switched on by soup autopilot for DPO, GRPO and pretrain

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#806]
## The defect

`training.use_flash_attn` and `training.use_liger` have exactly one training consumer, `SFTTrainerWrapper._setup_transformers`:

- `trainer/sft.py:1293` applies the Liger patch before model load, and `sft.py:882` sets `use_liger_kernel` on `TrainingArguments`;

- `trainer/sft.py:1351-1356` sets `attn_implementation` from `use_flash_attn`.

No other trainer, and not SFT's vision, audio or streaming branches, references either field or sets `attn_implementation` (grep over `trainer/` outside `sft.py` and `mlx_sft.py` finds nothing). Yet:

1. **`commands/train.py:1192-1214` validates both for every task** and exits 1 when the package is missing. A DPO run can be refused for a missing `liger-kernel` that DPO would never use.

2. **`soup autopilot` enables both** on Ampere or newer (`autopilot/decisions.py:344-346`) whatever the task, and its goal map sends `alignment` to DPO, `reasoning` to GRPO and `domain-adapt` to pretrain (`decisions.py:15-17`).

## Reproduction

`build_soup_config` run on this machine (compute capability 12.0) with a preference JSONL:

```

task alignment -> dpo use_flash_attn= True use_liger= True

task reasoning -> grpo use_flash_attn= True use_liger= True

task domain-adapt -> pretrain use_flash_attn= True use_liger= True

```

On the same machine neither package is installed, so the validators `soup train` runs for any task return:

```

liger installed: False | flash-attn: None

['liger-kernel is not installed. Install it with: pip install "soup-cli[liger]"']

['FlashAttention is not available. Install it with: pip install flash-attn --no-build-isolation']

```

So the zero-config path writes a DPO config that `soup train` refuses, for two settings DPO ignores. Installing both packages makes the run start, still without either kernel. Hand-written configs are accepted on every task too:

```

ACCEPT dpo  ACCEPT grpo  ACCEPT pretrain  ACCEPT kto  ACCEPT orpo  ACCEPT embedding  ACCEPT distill

  {'use_flash_attn': True, 'use_liger': True, 'rope_scaling_type': 'dynamic', ...}

```

`docs/performance-and-quantization.md:806-808` presents both with no task scope.

## Same shape: rope_scaling_type

`rope_scaling_type` is also applied only in `sft.py:1368` and validated for every task at `commands/train.py:1229`. Whether it works on SFT is a separate bug; here it only shares the SFT-only shape.

## Same shape, narrower: reasoning_effort / train_on_eot

`_validate_reasoning_effort_task_gate` (`config/schema.py:4780`) exists to stop exactly this, but its allowlist is wider than the consumers. Both fields are read only inside `build_format_row` (`data/sft_format.py:59-63`), called by SFT and distill. The allowlist also admits `pretrain`, `classifier`, `reranker` and `cross_encoder`, none of which calls `build_format_row`:

```

ACCEPT pretrain       {'reasoning_effort': 'high', 'train_on_eot': True}

ACCEPT classifier     {'reasoning_effort': 'high', 'train_on_eot': True}

ACCEPT reranker       {'reasoning_effort': 'high', 'train_on_eot': True}

ACCEPT cross_encoder  {'reasoning_effort': 'high', 'train_on_eot': True}

REJECT dpo  training.reasoning_effort='high' requires task in ['classifier', 'cross_encoder', 'distill', 'pretrain', 'reranker', 'sft']

```

The fix there is to shrink the set to `{"sft", "distill"}`, or to wire the four.

## Why it matters

These are the two most common speed flags, and autopilot sets them itself, so users do not inspect them. On DPO/GRPO a user pays for the install and the validation and gets eager attention with no fused kernels, while the config and the autopilot panel (`commands/autopilot.py:147-148`, "Flash Attention: True") say otherwise.

## Proposed fix

Either wire both where the model is loaded (an `attn_implementation` kwarg on each trainer's `from_pretrained` call and the pre-load Liger patch; both are small), or refuse them at config load on tasks that do not read them, as #745 proposes for `loraplus_lr_ratio`. In either case:

- move the `commands/train.py` availability checks behind the same task condition, so a flag that will be ignored cannot fail a run;

- make `decide_performance_flags` task-aware, so autopilot never writes a flag the chosen task ignores.

## Acceptance criteria

- [ ] A test that builds the DPO (and GRPO) model on a tiny CPU model with `use_flash_attn: true` and a stubbed `get_attn_implementation` returning `"sdpa"`, and asserts the kwarg reaches `from_pretrained`. Fails today; or config load rejects the flag for DPO.

- [ ] `build_soup_config(goal="alignment")` on a compute capability 8.0+ profile never emits a flag the resulting task ignores.

- [ ] `reasoning_effort` / `train_on_eot` on `task: pretrain` are applied or rejected at load.

- [ ] `soup train` does not exit 1 over a missing package for a flag the task does not use.

---

**Related, filed the same day:**

- #793 — rope_scaling_type is a no-op on transformers 5: inv_freq and logits are byte-identical after "Long-context ...
- #798 — MoE flags only work on SFT: moe_expert_quant, train_router_only, moe_lora and moe_aux_loss_coeff are accept...
- #801 — kernel_auto_compose times one unchanged model four times, prints the fastest name, and applies nothing
- #802 — loss_spike_recovery and grad_accum_auto_tune reach the training callback on SFT only: 15 trainers build it ...
- #803 — lora.rank_pattern / alpha_pattern reach 2 of 20 LoRA construction sites: DPO and KTO train q_proj at r=16 w...
- #804 — data.chat_template is applied by SFT and distill only: DPO renders conversational rows with the model's shi...

**Comment by AmirF194** [comment]
Taking this one. Scoping the fix to sft.py's `_setup_transformers` (and tts.py, which inherits it via `super()`), the only two places that actually read `use_flash_attn`/`use_liger`. train.py's validation gate and autopilot's auto-enable both key off that same set now.

**Comment by MakazhanAlpamys** [comment]
Partial progress: **#983 by @AmirF194** merged as `09e1a137`. **This issue stays open** — one acceptance criterion is deliberately left.

`training.use_flash_attn` and `training.use_liger` are now rejected at config load outside `{sft, tts}`, the only trainers that read either flag. The allowlist is `{sft, tts}` and not `{sft}` because `TTSTrainerWrapper` inherits `sft.py::_setup_transformers` — a scope-by-task-name reading of this issue lands on `sft` alone and silently breaks every TTS config that sets either flag.

`soup autopilot` stops auto-enabling both for DPO/GRPO/pretrain on Ampere or newer, where it previously emitted a config Soup's own validation would then refuse. `soup migrate --from axolotl` and `generate_config`'s flat decisions path got the same scoping. The migrate half was the one I re-ran rather than trusting the diff, on an axolotl config with `rl: dpo` and `flash_attention: true`:

```
before : migrate emits use_flash_attn=True  -> REFUSED at load
after  : migrate emits use_flash_attn=None  -> the config LOADS
```

The tempting fix there is to make the loader lenient; instead the migrator stopped carrying a flag the target task never reads, which keeps the new rejection strict.

**Still open:** the "same shape, narrower" note in this issue about `reasoning_effort` / `train_on_eot`. That is a different function with a different consumer (`build_format_row`, sft.py + distill.py), and whether the TTS formatting path calls it the way sft's text path does has not been checked. @AmirF194 flagged it as unverified rather than guessing, which is why it is still listed here.

Available to anyone — comment if you want it.

**Comment by MakazhanAlpamys** [comment]
I've removed the assignee and the `claimed` label so they match my 09-15 note: #983 by @AmirF194 shipped the flash-attn/liger half, and the `reasoning_effort` / `train_on_eot` remainder is open to anyone. On `main`, `_validate_reasoning_effort_task_gate` still admits pretrain, classifier, reranker and cross_encoder, and none of them reaches `build_format_row`, whose only trainer callers are sft and distill. The 09-13 ruling on #808 treats #793-#806 as bugs rather than staged fields, so this can be refused at load directly, with no warning release.
